ledger-service/http-json/src/itlib/scala/http/AbstractWebsocketServiceIntegrationTest.scala

```diff
@@ -916,6 +953,173 @@ abstract class AbstractWebsocketServiceIntegrationTest
       }
   }
 
+  "Per-query offsets should work as expected" in withHttpService { (uri, _, _) =>
+    val dslSyntax = Consume.syntax[JsValue]
+    import dslSyntax._
+    import spray.json._
+    def createIouCommand(currency: String): String =
+      s"""{
+           |  "templateId": "Iou:Iou",
+           |  "payload": {
+           |    "observers": [],
+           |    "issuer": "Alice",
+           |    "amount": "999.99",
+           |    "currency": "$currency",
+           |    "owner": "Alice"
+           |  }
+           |}""".stripMargin
+    def createIou(currency: String): Future[Assertion] =
+      HttpServiceTestFixture
+        .postJsonStringRequest(
+          uri.withPath(Uri.Path("/v1/create")),
+          createIouCommand(currency),
+          headersWithAuth,
+        )
+        .map(_._1 shouldBe a[StatusCodes.Success])
+    def contractsQuery(currency: String): String =
+      s"""{"templateIds":["Iou:Iou"], "query":{"currency":"$currency"}}"""
+    def contractsQueryWithOffset(offset: domain.Offset, currency: String): String =
+      s"""{"templateIds":["Iou:Iou"], "query":{"currency":"$currency"}, "query_offset":"${offset.unwrap}"}"""
+    def contracts(currency: String, offset: Option[domain.Offset]): String =
+      offset.fold(contractsQuery(currency))(contractsQueryWithOffset(_, currency))
+    def ledgerEnd(expectedContracts: Int): Future[domain.Offset] = {
+      def go(killSwitch: UniqueKillSwitch): Sink[JsValue, Future[domain.Offset]] =
+        Consume.interpret(
+          for {
+            _ <- readAcsN(expectedContracts)
+            ContractDelta(Vector(), _, Some(offset)) <- readOne
+            _ = killSwitch.shutdown()
+            _ <- drain
+          } yield offset
+        )
+      val (killSwitch, source) =
+        singleClientQueryStream(jwt, uri, """{"templateIds":["Iou:Iou"]}""")
+          .viaMat(KillSwitches.single)(Keep.right)
+          .preMaterialize()
+      source.via(parseResp).runWith(go(killSwitch))
+    }
+    def test(
+        clue: String,
+        expectedAcsSize: Int,
+        end: domain.Offset,
+        queryFrom: Option[domain.Offset],
+        eurFrom: Option[domain.Offset],
+        usdFrom: Option[domain.Offset],
+        expected: Map[String, Int],
+    ): Future[Assertion] = {
+      val expectLivenessMarker = expectedAcsSize != 0
+      def go(killSwitch: UniqueKillSwitch): Sink[JsValue, Future[Assertion]] = {
+        Consume.interpret(
+          for {
+            acs <- readAcsN(expectedAcsSize)
+            contracts <- updateAcs(acs, expectLivenessMarker, end)
+            result = contracts
+              .map(_._2.asJsObject.fields("currency").asInstanceOf[JsString].value)
+              .groupBy(identity)
+              .map { case (k, vs) => k -> vs.size }
+            // If this fails with a match error, there is something else lingering in the stream
+            ContractDelta(Vector(), _, _) <- readOne
```

Comment:
```suggestion
            ContractDelta(Vector(), Vector(), Some(_)) <- readOne
```
afaict you are reading the heartbeat here (as made explicit by my suggestion). We send heartbeats every 5 seconds. With 6 tests, this gives you roughly a 30s delay at worst. Probably not the only source of slowness but maybe one.

Easy fix is to just drop this but it does unfortunately mean that you are more risky to drop something on the floor. Alternatively, you can increase the heartbeating frequency to reduce the wait.

Reply:
Removing this did cut 30 seconds off. If we drop something the test should come back negative, so that should be ok. Thanks!